core/src/main/scala/org/http4s/Service.scala

```diff
@@ -2,32 +2,372 @@ package org.http4s
 
 import scalaz.Kleisli
 import scalaz.concurrent.Task
-import scalaz.syntax.kleisli._
+
+final case class Service[A, B](run: PartialFunction[A, Task[B]]) {
+  import Service._
+
+  def apply(a: A)(implicit ev: Task[Response] =:= Task[B]): Task[B] =
+    applyOrElse(a, (_: A) => ev(HttpService.notFound))
```

Comment:
I've always had mixed feelings about evidence pattern. It seems to work very well, but also feels wrong, I don't know why. I'd love to get others take on them, specifically how this pattern would be accepted in the Haskell world (or not...). It feels like a very Haskell'ish pattern, but not quite.